src/VisualStudio/Core/Def/Implementation/ProjectSystem/VisualStudioWorkspaceImpl.ProjectPropertyStorage.cs

```diff
@@ -0,0 +1,58 @@
+﻿// Copyright (c) Microsoft.  All Rights Reserved.  Licensed under the Apache License, Version 2.0.  See License.txt in the project root for license information.
+
+using System;
+using System.Linq;
+using EnvDTE;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Roslyn.Utilities;
+
+namespace Microsoft.VisualStudio.LanguageServices.Implementation.ProjectSystem
+{
+    internal partial class VisualStudioWorkspaceImpl
+    {
+        private abstract class ProjectPropertyStorage
+        {
+            public static ProjectPropertyStorage Create(Project project, IServiceProvider serviceProvider)
+            {
+                var solution = (IVsSolution)serviceProvider.GetService(typeof(SVsSolution));
+                solution.GetProjectOfUniqueName(project.UniqueName, out var hierarchy);
+
+                return hierarchy.IsCapabilityMatch("CPS")
+                    ? new BuildPropertyStorage((IVsBuildPropertyStorage)hierarchy)
+                    : new PerConfigurationPropertyStorage(project.ConfigurationManager) as ProjectPropertyStorage;
+            }
+
+            public abstract void SetProperty(string buildPropertyName, string configurationPropertyName, string value);
+
+            private sealed class BuildPropertyStorage : ProjectPropertyStorage
+            {
+                private readonly IVsBuildPropertyStorage propertyStorage;
+
+                public BuildPropertyStorage(IVsBuildPropertyStorage propertyStorage)
+                    => this.propertyStorage = propertyStorage;
+
+                public override void SetProperty(string buildPropertyName, string configurationPropertyName, string value)
+                {
+                    propertyStorage.SetPropertyValue(buildPropertyName, null, (uint)_PersistStorageType.PST_PROJECT_FILE, value);
+                }
+            }
+
+            private sealed class PerConfigurationPropertyStorage : ProjectPropertyStorage
+            {
+                private readonly ConfigurationManager configurationManager;
+
+                public PerConfigurationPropertyStorage(ConfigurationManager configurationManager)
+                    => this.configurationManager = configurationManager;
+
+                public override void SetProperty(string buildPropertyName, string configurationPropertyName, string value)
+                {
+                    foreach (Configuration configuration in configurationManager)
+                    {
+                        configuration.Properties.Cast<Property>().First(property => property.Name == configurationPropertyName).Value = value;
```

Comment:
1. consider splitting over multiple lines.  i.e. do the 'get' first then have another statement to do the assignment.
2. what happens if there is no property with that name?  Perhaps FirstOrDefault so we don't crash?

Reply:
@jasonmalinowski ?

Reply:
There should always be a property if you pass in the right name. If it's not found, it's probably a bug in the implementation. So crashing with "code fix encountered a problem" at that point would seem appropriate so we can investigate and fix it.

Reply:
> There should always be a property if you pass in the right name

Can you point to whatever documentation or other source leads you to believe that htat's the case?  Given our history with project systems, i can say pretty reasonably that if it isn't documented somewhere, it may absolutely not abide by that sort of assumption either now, or at some point in the future.

Reply:
If this is doc'ed somewhere always being available, then that's ok with me.  If that's the case, can we add a comment with an appropriate link which then explains why it's ok to be able to assume this?  Thanks!

Reply:
my understanding is this should enumrate all properties that are in `VSLangProj80.CSharpProjectConfigurationProperties3`. @jasonmalinowski can correct me if i'm wrong